File: mask-ts/src/core/fpe.ts
/**
 * Deterministic Pseudonymization (DP) token generation using NIST SP 800-38G FF1.
 */

import * as crypto from 'crypto';
import { config } from '../config';
import { getKeyProvider } from './key_provider';
import { MaskSecurityError } from './exceptions';
import { FF1 } from './ff1';
import {
  FIRST_NAMES as _BIJECTIVE_NAMES,
  CONNECTORS as _BIJECTIVE_CONNECTORS,
  SURNAME_ROOTS as _BIJECTIVE_ROOTS,
  SURNAME_SUFFIXES as _BIJECTIVE_SUFFIXES,
  SYLLABLES as _BIJECTIVE_SYLLABLES
} from './synthesisLibrary';

let _masterKey: Buffer | null = null;

async function _getMasterKey(): Promise<Buffer> {
  if (_masterKey === null) {
    const provider = getKeyProvider();
    let raw = await provider.getMasterKey();
    
    if (!raw) {
      raw = await provider.getEncryptionKey() || "";
    }

    if (!raw) {
      if (config.MASK_DEV_MODE) {
        raw = crypto.randomBytes(32).toString('hex');
        process.env.MASK_MASTER_KEY = raw;
      } else {
        throw new MaskSecurityError("MASK_MASTER_KEY not set.");
      }
    }
    _masterKey = Buffer.from(raw, 'utf-8');
  }
  return _masterKey;
}

export function resetMasterKey(): void {
  _masterKey = null;
}

async function _getAesKey(): Promise<Buffer> {
  // Salt the derivation with the tenant ID to guarantee per-tenant FF1
  // uniqueness — two tenants with the same plaintext must never produce
  // the same FPE token (cross-tenant collision prevention).
  const masterKey = await _getMasterKey();
  return crypto.createHmac('sha256', masterKey).update(config.MASK_TENANT_ID, 'utf-8').digest();
}

const _EMAIL_RE = /^[^@\s]+@[^@\s]+\.[^@\s]+$/;
const _PHONE_RE = /(?<!\d)(?:\+?1?[\s\-.]?\(?\d{3}\)?[\s\-.]?\d{3}[\s\-.]?\d{4}|\d{3}[\s\-.]?\d{4}|\+\d{2,3}[\s\-.]?\d{3}[\s\-.]?\d{3}[\s\-.]?\d{3,4})(?!\d)/;
const _SSN_RE = /^\d{3}-\d{2}-\d{4}$/;
const _CC_RE = /^(?:\d{4}[ \-]?){3}\d{4}$/;
const _ROUTING_RE = /^\d{9}$/;
const _ES_ID_RE = /^(?:\d{8}[A-Z]|[XYZ]\d{7}[A-Z])$/;
const _IBAN_RE = /^[A-Z]{2}\d{2}[A-Z0-9]{4,30}$/;

async function _hmacHex(plaintext: string, n: number = 8): Promise<string> {
  const masterKey = await _getMasterKey();
  const digest = crypto.createHmac('sha256', masterKey).update(plaintext, 'utf-8').digest('hex');
  return digest.slice(0, n);
}

// ── Bijective Synthesis Engine ─────────────────────────────────────────────

async function _getBijectiveTweak(): Promise<Buffer> {
  /**
   * Derive the FF1 tweak deterministically from the tenant ID.
   *
   * IMPORTANT: The tweak is intentionally time-independent. Historical use of
   * MASK_SALT_ROTATION (MONTHLY/YEARLY) caused permanent data loss when the
   * calendar rolled over because old tokens could no longer be re-derived.
   * Use MASK_KEYRING for key rotation instead; MASK_SALT_ROTATION is now a
   * no-op and will emit a console.warn if set to a non-NONE value.
   */
  if (config.MASK_SALT_ROTATION !== 'NONE') {
    console.warn(
      `[mask] MASK_SALT_ROTATION=${config.MASK_SALT_ROTATION} is deprecated and ignored. ` +
      'Time-based tweaks caused permanent data loss on month/year rollovers. ' +
      'Use MASK_KEYRING for key rotation instead.'
    );
  }
  const masterKey = await _getMasterKey();
  return crypto.createHmac('sha256', masterKey).update(config.MASK_TENANT_ID, 'utf-8').digest();
}

async function _encryptBijectiveFF1(text: string): Promise<bigint> {
  const canonical = text.toLowerCase().trim();
  const hash = crypto.createHash('sha256').update(canonical, 'utf-8').digest();
  // Hash to 64-bit int, then to 20-digit string
  const inputInt = hash.readBigUInt64BE(0);
  const inputStr = inputInt.toString().padStart(20, '0');
  
  const aesKey = await _getAesKey();
  const tweak = await _getBijectiveTweak();
  const engine = new FF1(aesKey, tweak, 10);
  
  const cipherStr = engine.encrypt(inputStr);
  return BigInt(cipherStr) % (2n ** 64n);
}

function _renderBijectivePerson(bits: bigint): string {
  const firstIdx = Number(bits & 0x7FFn);          // 11 bits (2048)
  const connIdx = Number((bits >> 11n) & 0x3Fn);    // 6 bits (64)
  const rootIdx = Number((bits >> 17n) & 0xFFFn);    // 12 bits (4096)
  const suffixIdx = Number((bits >> 29n) & 0x1FFn); // 9 bits (512)
  const tag = Number((bits >> 38n) & 0x3FFFn);      // 14 bits (16384)
  const formatIdx = Number((bits >> 52n) & 0xFn);   // 4 bits (16)

  const first = _BIJECTIVE_NAMES[firstIdx % _BIJECTIVE_NAMES.length];
  const conn = _BIJECTIVE_CONNECTORS[connIdx % _BIJECTIVE_CONNECTORS.length];
  const root = _BIJECTIVE_ROOTS[rootIdx % _BIJECTIVE_ROOTS.length];
  const suffix = _BIJECTIVE_SUFFIXES[suffixIdx % _BIJECTIVE_SUFFIXES.length];
  const surname = `${root}${suffix}`;
  const numeric = tag % 10000;

  const paddedNumeric = numeric.toString().padStart(4, '0');

  if (formatIdx === 0) return `${first} ${conn} ${surname}-${paddedNumeric}`;
  if (formatIdx === 1) return `${surname}, ${first}-${paddedNumeric}`;
  if (formatIdx === 2) return `${first[0]}. ${surname}-${paddedNumeric}`;
  if (formatIdx === 3) return `${first} ${surname}-${paddedNumeric}`;
  
  return `${first} ${surname}-${paddedNumeric}`;
}

function _renderBijectiveLocation(bits: bigint): string {
  const s1 = Number(bits & 0x3FFn);
  const s2 = Number((bits >> 10n) & 0x3FFn);
  const s3 = Number((bits >> 20n) & 0x3FFn);
  const tag = Number((bits >> 30n) & 0xFFFn);

  const city = `${_BIJECTIVE_SYLLABLES[s1 % 1000]}${_BIJECTIVE_SYLLABLES[s2 % 1000].toLowerCase()}${_BIJECTIVE_SYLLABLES[s3 % 1000].toLowerCase()}`;
  return `${city}-${tag.toString().padStart(3, '0')}`;
}

function _computeLuhnDigit(partialNum: string): string {
    const digits = partialNum.split("").map(Number);
    let sum = 0;
    let shouldDouble = true; 
    for (let i = digits.length - 1; i >= 0; i--) {
        let digit = digits[i];
        if (shouldDouble) {
            digit *= 2;
            if (digit > 9) digit -= 9;
        }
        sum += digit;
        shouldDouble = !shouldDouble;
    }
    return ((10 - (sum % 10)) % 10).toString();
}

function _computeEsIdCheck(num: number): string {
  return "TRWAGMYFPDXBNJZSQVHLCKE"[num % 23];
}

function _stripCcSeparators(text: string): string {
  return text.replace(/[\s\-]/g, '');
}

export async function generateDPToken(rawText: string, entityType: string = 'UNKNOWN'): Promise<string> {
  const text = rawText.trim();
  let type = (entityType || "UNKNOWN").toUpperCase();

  if (type === "UNKNOWN") {
    if (_EMAIL_RE.test(text)) type = "EMAIL_ADDRESS";
    else if (_SSN_RE.test(text)) type = "US_SSN";
    else if (_CC_RE.test(text)) type = "CREDIT_CARD";
    else if (_ROUTING_RE.test(text)) type = "US_ROUTING_NUMBER";
    else if (_ES_ID_RE.test(text)) type = "ES_ID";
    else if (_IBAN_RE.test(text)) type = "INTL_BANK_IBAN";
    else if (_PHONE_RE.test(text)) type = "PHONE_NUMBER";
  }

  if (type === "EMAIL_ADDRESS" || type === "EMAIL_ADDR") {
    const parts = text.split("@");
    const domain = parts.length === 2 ? parts[1] : "email.com";
    return `tkn-${await _hmacHex(text)}@${domain}`;
  }

  if (type === "PHONE_NUMBER" || type === "PHONE_NUM" || type === "PHONE_NUM_INTL") {
    const m = text.match(/^\+([1-9]\d{0,3})/);
    const cc = m ? m[1] : "1";
    const digits = text.replace(/\D/g, "");
    if (digits.length >= 7) {
      const last7 = digits.slice(-7);
      const engine = new FF1(await _getAesKey(), Buffer.from("PHONE"), 10);
      const enc = engine.encrypt(last7);
      return `+${cc}-555-${enc}`;
    }
  }

  if (type === "US_SSN") {
    const digits = text.replace(/-/g, "");
    if (digits.length === 9) {
      const engine = new FF1(await _getAesKey(), Buffer.from("US_SSN"), 10);
      const enc = engine.encrypt(digits);
      return `${enc.slice(0,3)}-${enc.slice(3,5)}-${enc.slice(5,9)}`;
    }
  }

  if (type === "CREDIT_CARD" || type === "CREDIT_CARD_NUMBER") {
    const digits = _stripCcSeparators(text);
    if (digits.length === 16) {
      const bin6 = digits.slice(0, 6);
      const last4 = digits.slice(12, 16);
      const middle6 = digits.slice(6, 12);
      
      const engine = new FF1(await _getAesKey(), Buffer.from("CREDIT_CARD"), 10);
      const encMiddle = engine.encrypt(middle6);
      
      const base15 = bin6 + encMiddle + last4.slice(0, 3);
      const checkDig = _computeLuhnDigit(base15);
      const full = bin6 + encMiddle + last4.slice(0, 3) + checkDig;
      return `${full.slice(0, 4)}-${full.slice(4, 8)}-${full.slice(8, 12)}-${full.slice(12, 16)}`;
    } else {
      const fallbackDigits = digits.padEnd(16, '0').slice(0, 16);
      const engine = new FF1(await _getAesKey(), Buffer.from("CREDIT_CARD"), 10);
      const encMiddle = engine.encrypt(fallbackDigits.slice(6, 12));
      const full = fallbackDigits.slice(0, 6) + encMiddle + fallbackDigits.slice(12);
      return `${full.slice(0, 4)}-${full.slice(4, 8)}-${full.slice(8, 12)}-${full.slice(12, 16)}`;
    }
  }

  if (type === "US_ROUTING_NUMBER" || type === "US_ABA_ROUTING") {
    if (text.length === 9 && /^\d+$/.test(text)) {
      const engine = new FF1(await _getAesKey(), Buffer.from("US_ROUTING"), 10);
      return engine.encrypt(text);
    }
  }

  if (type === "INTL_BANK_IBAN" || type === "IBAN_CODE") {
    const countryCode = (text.length >= 2 && /[a-zA-Z]{2}/.test(text.slice(0, 2))) ? text.slice(0, 2).toUpperCase() : "US";
    return `${countryCode}00${(await _hmacHex(text, 8)).toUpperCase()}`;
  }

  if (type === "ES_ID" || type === "ES_DNI") {
    let digits = text.toUpperCase().replace(/[A-Z]/g, "");
    if (digits) {
      digits = digits.padStart(8, "0");
      const engine = new FF1(await _getAesKey(), Buffer.from("ES_ID"), 10);
      const enc = engine.encrypt(digits.slice(-5));
      const tokenDigits = `000${enc}`;
      return tokenDigits + _computeEsIdCheck(parseInt(tokenDigits, 10));
    }
  }

  if (type === "PERSON" || type === "PERSON_NAME") {
      if (config.MASK_BIJECTIVE_MODE) {
        const cipherBits = await _encryptBijectiveFF1(text);
        return _renderBijectivePerson(cipherBits);
      }
      return `[TKN-PERSON-${await _hmacHex(text)}]`;
  }
  if (type === "LOCATION" || type === "PHYS_ADDRESS") {
      if (config.MASK_BIJECTIVE_MODE) {
        const cipherBits = await _encryptBijectiveFF1(text);
        return _renderBijectiveLocation(cipherBits);
      }
      return `[TKN-LOC-${await _hmacHex(text)}]`;
  }
  if (type === "ORGANIZATION") {
      return `[TKN-ORG-${await _hmacHex(text)}]`;
  }

  return `[TKN-${await _hmacHex(text)}]`;
}

export const generateFPEToken = generateDPToken;

export * from './fpe_utils';

File: mask-python/mask_privacy/core/fpe.py
import os
import hmac
import hashlib
import re
import logging
from typing import Optional, List

from mask_privacy import config
from mask_privacy.core.ff1 import FF1
from mask_privacy.core.synthesis_library import (
    FIRST_NAMES as _BIJECTIVE_NAMES,
    CONNECTORS as _BIJECTIVE_CONNECTORS,
    SURNAME_ROOTS as _BIJECTIVE_ROOTS,
    SURNAME_SUFFIXES as _BIJECTIVE_SUFFIXES,
    SYLLABLES as _BIJECTIVE_SYLLABLES
)

logger = logging.getLogger("mask.fpe")

# Master key management

_master_key: Optional[bytes] = None

def _get_master_key() -> bytes:
    """Return the HMAC master key, lazily initialised from the key provider."""
    global _master_key
    if _master_key is None:
        from mask_privacy.core.key_provider import get_key_provider
        raw = get_key_provider().get_master_key() or ""
        if not raw:
            import secrets
            raw = secrets.token_hex(32)
            os.environ["MASK_MASTER_KEY"] = raw
            if not config.MASK_DEV_MODE:
                from mask_privacy.core.exceptions import MaskSecurityError
                raise MaskSecurityError("MASK_MASTER_KEY not set.")
        _master_key = raw.encode("utf-8")
    return _master_key

def reset_master_key() -> None:
    """Clear the cached master key."""
    global _master_key
    _master_key = None

def _get_aes_key() -> bytes:
    # Ensure exactly 32 bytes for FF1 AES-256
    return hashlib.sha256(_get_master_key()).digest()

# Regex Detectors
_EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")
_PHONE_RE = re.compile(r"^\+?1?[\s\-.]?\(?\d{3}\)?[\s\-.]?\d{3}[\s\-.]?\d{4}$|^\d{3}[\s\-.]?\d{4}$|^\+\d{2,3}[\s\-.]?\d{3}[\s\-.]?\d{3}[\s\-.]?\d{3,4}$")
_SSN_RE   = re.compile(r"^\d{3}-\d{2}-\d{4}$")
_CC_RE    = re.compile(r"^(?:\d{4}[ \-]?){3}\d{4}$")
_ROUTING_RE = re.compile(r"^\d{9}$")
_ES_ID_RE = re.compile(r"^(?:\d{8}[A-Z]|[XYZ]\d{7}[A-Z])$")
_IBAN_RE = re.compile(r"^[A-Z]{2}\d{2}[A-Z0-9]{4,30}$")

def _hmac_hex(plaintext: str, n: int = 8) -> str:
    digest = hmac.new(_get_master_key(), plaintext.encode("utf-8"), hashlib.sha256).hexdigest()
    return digest[:n]


# ── Bijective Synthesis Engine ─────────────────────────────────

def _get_bijective_tweak() -> bytes:
    base = config.MASK_TENANT_ID.encode("utf-8")
    if config.MASK_SALT_ROTATION != "NONE":
        import datetime
        now = datetime.datetime.now()
        if config.MASK_SALT_ROTATION == "MONTHLY":
            base += f"-{now.year}-{now.month}".encode("utf-8")
        elif config.MASK_SALT_ROTATION == "YEARLY":
            base += f"-{now.year}".encode("utf-8")
    return hmac.new(_get_master_key(), base, hashlib.sha256).digest()

def _encrypt_bijective_ff1(text: str) -> int:
    canonical = text.lower().strip()
    # Hash to 64-bit int, then to 20-digit string
    input_str = str(int.from_bytes(hashlib.sha256(canonical.encode()).digest()[:8], "big")).zfill(20)
    engine = FF1(_get_aes_key(), _get_bijective_tweak(), 10)
    cipher_str = engine.encrypt(input_str)
    return int(cipher_str) % (2**64)

def _render_bijective_person(bits: int) -> str:
    # Bit allocation: First(11) + Conn(6) + Root(12) + Suffix(9) + Tag(14) = 52 bits
    first_idx = bits & 0x7FF
    conn_idx = (bits >> 11) & 0x3F
    root_idx = (bits >> 17) & 0xFFF
    suffix_idx = (bits >> 29) & 0x1FF
    tag = (bits >> 38) & 0x3FFF
    format_idx = (bits >> 52) & 0xF
    
    first = _BIJECTIVE_NAMES[first_idx % len(_BIJECTIVE_NAMES)]
    conn = _BIJECTIVE_CONNECTORS[conn_idx % len(_BIJECTIVE_CONNECTORS)]
    root = _BIJECTIVE_ROOTS[root_idx % len(_BIJECTIVE_ROOTS)]
    suffix = _BIJECTIVE_SUFFIXES[suffix_idx % len(_BIJECTIVE_SUFFIXES)]
    numeric = tag % 10000
    
    surname = f"{root}{suffix}"
    if format_idx == 0: return f"{first} {conn} {surname}-{numeric:04d}"
    if format_idx == 1: return f"{surname}, {first}-{numeric:04d}"
    if format_idx == 2: return f"{first[0]}. {surname}-{numeric:04d}"
    return f"{first} {surname}-{numeric:04d}"

def _render_bijective_location(bits: int) -> str:
    s1, s2, s3 = bits & 0x3FF, (bits >> 10) & 0x3FF, (bits >> 20) & 0x3FF
    tag = (bits >> 30) & 0xFFF
    city = f"{_BIJECTIVE_SYLLABLES[s1 % 1000]}{_BIJECTIVE_SYLLABLES[s2 % 1000].lower()}{_BIJECTIVE_SYLLABLES[s3 % 1000].lower()}"
    return f"{city}-{tag:03d}"

# Formatting & Luhn
def _compute_luhn_digit(partial_num: str) -> str:
    digits = [int(x) for x in partial_num]
    sum_ = 0
    should_double = True
    for digit in reversed(digits):
        if should_double:
            digit *= 2
            if digit > 9: digit -= 9
        sum_ += digit
        should_double = not should_double
    return str((10 - (sum_ % 10)) % 10)

def _compute_es_id_check(num: int) -> str:
    return "TRWAGMYFPDXBNJZSQVHLCKE"[num % 23]

def _strip_cc_separators(text: str) -> str:
    return re.sub(r"[\s\-]", "", text)

# Public API
def generate_dp_token(raw_text: str, entity_type: str = "UNKNOWN") -> str:
    """Generate a deterministic, format-preserving pseudonymization token using NIST FF1.
    """
    text = raw_text.strip()
    type_ = (entity_type or "UNKNOWN").upper()

    if type_ == "UNKNOWN":
        if _EMAIL_RE.match(text): type_ = "EMAIL_ADDRESS"
        elif _SSN_RE.match(text): type_ = "US_SSN"
        elif _CC_RE.match(text): type_ = "CREDIT_CARD"
        elif _ROUTING_RE.match(text): type_ = "US_ROUTING_NUMBER"
        elif _ES_ID_RE.match(text): type_ = "ES_ID"
        elif _IBAN_RE.match(text): type_ = "INTL_BANK_IBAN"
        elif _PHONE_RE.match(text): type_ = "PHONE_NUMBER"

    if type_ in ("EMAIL_ADDRESS", "EMAIL_ADDR"):
        parts = text.split("@", 1)
        domain = parts[1] if len(parts) == 2 else "email.com"
        return f"tkn-{_hmac_hex(text)}@{domain}"

    if type_ == "US_SSN":
        # Pure FF1 FPE over all 9 digits
        digits = text.replace("-", "")
        if len(digits) == 9 and digits.isdigit():
            engine = FF1(_get_aes_key(), b"US_SSN", 10)
            enc = engine.encrypt(digits)
            return f"{enc[:3]}-{enc[3:5]}-{enc[5:9]}"
        
    if type_ in ("CREDIT_CARD", "CREDIT_CARD_NUMBER"):
        digits = _strip_cc_separators(text)
        if len(digits) == 16:
            bin6 = digits[:6]
            last4 = digits[12:16]
            middle6 = digits[6:12]
            engine = FF1(_get_aes_key(), b"CREDIT_CARD", 10)
            enc_middle = engine.encrypt(middle6)
            base15 = bin6 + enc_middle + last4[:3]  # 15 digits for Luhn input
            check = _compute_luhn_digit(base15)
            full = bin6 + enc_middle + last4[:3] + check
            return f"{full[:4]}-{full[4:8]}-{full[8:12]}-{full[12:16]}"
        else:
            # Fallback length
            fallback_digits = digits.ljust(16, "0")[:16]
            engine = FF1(_get_aes_key(), b"CREDIT_CARD", 10)
            enc_middle = engine.encrypt(fallback_digits[6:12])
            full = fallback_digits[:6] + enc_middle + fallback_digits[12:]
            return f"{full[:4]}-{full[4:8]}-{full[8:12]}-{full[12:16]}"

    if type_ in ("US_ROUTING_NUMBER", "US_ABA_ROUTING"):
        if len(text) == 9 and text.isdigit():
            engine = FF1(_get_aes_key(), b"US_ROUTING", 10)
            enc = engine.encrypt(text)
            return enc

    if type_ in ("INTL_BANK_IBAN", "IBAN_CODE"):
        country = text[:2].upper() if len(text) >= 2 and text[:2].isalpha() else "US"
        return f"{country}00{_hmac_hex(text, n=8).upper()}"

    if type_ in ("ES_ID", "ES_DNI"):
        digits = re.sub(r"[A-Z]", "", text.upper())
        if digits:
            digits = digits.zfill(8)
            engine = FF1(_get_aes_key(), b"ES_ID", 10)
            enc = engine.encrypt(digits[-5:])
            token_digits = f"000{enc}"
            return token_digits + _compute_es_id_check(int(token_digits))

    if type_ in ("PHONE_NUMBER", "PHONE_NUM", "PHONE_NUM_INTL"):
        m = re.match(r"^\+([1-9]\d{0,3})", text)
        cc = m.group(1) if m else "1"
        digits = re.sub(r"\D", "", text)
        if len(digits) >= 7:
            last7 = digits[-7:]
            engine = FF1(_get_aes_key(), b"PHONE", 10)
            enc = engine.encrypt(last7)
            return f"+{cc}-555-{enc}"

    if type_ in ("PERSON", "PERSON_NAME"):
        if config.MASK_BIJECTIVE_MODE:
            cipher_bits = _encrypt_bijective_ff1(text)
            return _render_bijective_person(cipher_bits)
        return f"[TKN-PERSON-{_hmac_hex(text)}]"

    if type_ in ("LOCATION", "PHYS_ADDRESS"):
        if config.MASK_BIJECTIVE_MODE:
            cipher_bits = _encrypt_bijective_ff1(text)
            return _render_bijective_location(cipher_bits)
        return f"[TKN-LOC-{_hmac_hex(text)}]"

    if type_ in ("ORGANIZATION",):
        return f"[TKN-ORG-{_hmac_hex(text)}]"

    return f"[TKN-{_hmac_hex(text)}]"

generate_fpe_token = generate_dp_token

TOKEN_PATTERN = re.compile(
    r"tkn-[a-f0-9]{8,64}@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}"
    r"|\+[1-9]\d{0,3}-555-\d{7}"
    r"|\d{3}-\d{2}-\d{4}"          
    r"|\d{4}-\d{4}-\d{4}-\d{4}"   
    r"|\b\d{9}\b"                       
    r"|\b000\d{5}[A-Z]\b"              
    r"|[A-Z]{2}00[A-F0-9]{4,16}"  
    r"|<(?:PER|LOC|ORG):[^>]+>"
    r"|\[TKN-[^\]]+\]"
    r"|\b[A-Z][a-zA-Z, ]+-[0-9]{3,4}\b"  
)

def looks_like_token(value: str) -> bool:
    v = value.strip()
    if v.startswith("tkn-") and "@" in v: return True
    if re.match(r"^\+[1-9]\d{0,3}-555-\d{7}$", v): return True
    if re.match(r"^\d{3}-\d{2}-\d{4}$", v): return True
    if re.match(r"^\d{4}-\d{4}-\d{4}-\d{4}$", v): return True
    if len(v) == 9 and v.isdigit(): return True
    if len(v) >= 8 and v[:2].isalpha() and v[2:4] == "00": return True
    if re.match(r"^000\d{5}[A-Z]$", v): return True
    if re.match(r"^<(PER|LOC|ORG):[^>]+>$", v): return True
    if v.startswith("[TKN-") and v.endswith("]"): return True
    if re.match(r"^[A-Z][a-zA-Z, ]+-[0-9]{3,4}$", v): return True
    return False
